src/cargo/core/compiler/custom_build.rs

```diff
@@ -405,7 +405,21 @@ fn build_work(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Job> {
                 },
                 true,
             )
-            .with_context(|| format!("failed to run custom build command for `{}`", pkg_descr));
+            .with_context(|| {
+                let mut build_error_context = format!("failed to run custom build command for `{}`", pkg_descr);
+
+                // If we're opting into backtraces, mention that build dependencies' backtraces can
+                // be improved by setting a higher debuginfo level.
+                if let Ok(show_backtraces) = std::env::var("RUST_BACKTRACE") {
+                    if show_backtraces != "0" {
+                        build_error_context.push_str("\n\
+                            note: To improve backtraces for build dependencies, make sure full debug info is turned on. \
+                            More details at https://doc.rust-lang.org/cargo/reference/profiles.html#build-dependencies");
```

Comment:
A couple suggestions here to make this a little more user friendly:

Can this print the environment variable the user needs to set? Looking up documentation and modifying `Cargo.toml` may be a bit onerous just to display a better backtrace. I think it can be done with something like:

```rust
// Outside the closure:
let env_profile_name = unit.profile.name.to_uppercase();
// ...
// Inside this error message:
format!("Set the CARGO_PROFILE_{env_profile_name}_BUILD_OVERRIDE_DEBUG=true environment variable to enable debug generation.\n")
```

The other suggestion is to only show this extra message if debug is actually turned off. Perhaps something like this:

```rust
// Outside the closure:
// I think this is roughly correct, but please scrutinize it.
let built_with_debug = cx
    .bcx
    .unit_graph
    .get(unit)
    .and_then(|deps| deps.iter().find(|dep| dep.unit.target == unit.target))
    .map(|dep| dep.unit.profile.debuginfo.is_turned_on())
    .unwrap_or(false);
// Inside the closure:
if !built_with_debug && show_backtraces != "0" {
   // ...
}
```

WDYT?

Reply:
Done, and done. Let me know if you like the wording.

Thanks for the snippet, as I'm sure I would have tried looking at `unit.profile.debuginfo` first and would have been confused as to why the message didn't appear when expected.